### PR TITLE
fix: mark FA2 auxiliary outputs non-differentiable

### LINKED ISSUE #1817: softmax_lse is silently not backpropogated
`softmax_lse` currently does not support backpropogation. This is fine, however the fact that it doesn't trigger any error makes it problematic IMO.
Suggested fix: Add something like
```
for d in args:
    torch._assert_async((d == 0).all())
```
to the backward functions in `flash_attn_interface.py` (what I don't like about this specific solution is that the error messages are not very informative).
Minimal example:
```
import torch
from flash_attn.flash_attn_interface import flash_attn_func

B, SEQ_LEN, H, D = 4, 2048, 16, 64
dtype = torch.bfloat16
device = 'cuda'

q = torch.randn(B, SEQ_LEN, H, D, device=device, dtype=dtype, requires_grad=True)
k = torch.randn(B, SEQ_LEN, H, D, device=device, dtype=dtype, requires_grad=True)
v = torch.randn(B, SEQ_LEN, H, D, device=device, dtype=dtype, requires_grad=True)

attn_output, softmax_lse_flash, _ = flash_attn_func(q=q, k=k, v=v, return_attn_probs=True)

loss = softmax_lse_flash.mean()
# loss = attn_output.mean() # this case is fine
loss.backward()

# prints zeroes, which is incorrect
print(q.grad.abs().max().item())
print(k.grad.abs().max().item())
print(v.grad.abs().max().item())
```

### PR DESCRIPTION
**Pull request by zupengwang**
## Summary

- mark `softmax_lse` and `S_dmask` as non-differentiable in all six FA2 autograd wrappers when `return_attn_probs=True`
- preserve the existing gradient path through the primary `out` tensor
- cover fixed-length and varlen APIs in QKV-packed, KV-packed, and separate Q/K/V forms

Fixes #1817.

## Motivation

The auxiliary outputs are intended for testing and their gradients are not implemented. Previously, a loss built only from `softmax_lse` silently materialized zero gradients for Q, K, and V. Marking both auxiliary outputs as non-differentiable makes that unsupported use fail explicitly while leaving normal output backpropagation unchanged.

Related work: #2405 independently marks the auxiliary outputs in `FlashAttnFunc` while adding `setup_context` support. This change applies the #1817 behavior consistently to all six current FA2 wrappers.

## Testing

- source build: `FLASH_ATTENTION_FORCE_BUILD=TRUE FLASH_ATTN_CUDA_ARCHS=80 MAX_JOBS=16 NVCC_THREADS=2 python -m pip install -e . --no-build-isolation --no-deps`
- regression: `python -m pytest -q -s tests/test_flash_attn.py::test_flash_attn_aux_outputs_are_non_differentiable` (`6 passed`)
- upstream PyTorch-reference helpers: 7 forward/backward cases across fixed/varlen, FP16/BF16, causal/non-causal, dropout `0.0/0.17`, and head dimensions `32/64/128/192/256`
- syntax and whitespace: `python -m py_compile flash_attn/flash_attn_interface.py tests/test_flash_attn.py` and `git diff --check`

Validated on an NVIDIA GeForce RTX 3090 (SM86) with PyTorch `2.13.0+cu130` and CUDA 13.0. Only one GPU was used because the change has no distributed semantics.

### CODE REVIEW

**Review by bvolpato — APPROVED** [review]
LGTM. Nice focused fix, consistent across all six wrappers, with primary `out` gradient path explicitly covered.

I ran all three fixed-length regression variants on CUDA, plus an existing zero-dropout forward/backward reference case. `py_compile` and `git diff --check` pass. I also reviewed varlen paths statically; local varlen execution was blocked by an older extension ABI, while the PR source-build run reports all six passing.
